api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSource.java

```diff
@@ -198,4 +201,103 @@ default int getOrdinal() {
      * @return the name of the configuration source
      */
     String getName();
+
+    /**
+     * Type of this config source. For backward compatibility returns {@link Type#PROPERTY} by default.
+     *
+     * @return source type
+     */
+    default Type getType() {
+        return Type.PROPERTY;
+    }
+
+    /**
+     * Whether the content of this config source can change externally.
+     * For backward compatibility returns {@code true} by default.
+     * <p>
+     * This information is required when the type is other than {@link Type#PROPERTY}.
+     * <p>
+     * Examples:
+     * <ul>
+     *     <li>Environment variables are not mutable</li>
+     *     <li>System properties may be considered mutable</li>
+     *     <li>Classpath resource from a jar file is not mutable</li>
+     *     <li>File based config source may be considered mutable</li>
+     * </ul>
+     *
+     * @return whether this is a mutable config source
+     */
+    default boolean isMutable() {
+        return true;
+    }
+
+    /**
+     * If this source {@link #getType()} is {@link Type#TREE}, this method is used to load the
+     * tree of the data.
+     *
+     * @return node content or empty optional if the underlying config source does not exist
+     */
+    default Optional<ConfigNode> load() {
```

Comment:
I think we should need this method if we update the following method `load(String key)` to state that if the `key` is null, it will load the whole tree. Otherwise, load the tree with the specified `root`. I would change the signature `load(String root)` while `null` means the whole root.

Reply:
I am against using `null` as a valid parameter value (anywhere, anytime). 
I can see two options here:
1. use empty string instead (as the root node does not have a name, so the key could be an empty string)
2. use a dedicated `Key` class that would have the concept of a root and use that